File: backend/routers/syntax.py
"""
Syntax Analysis Router
API endpoints for constituency and dependency parsing
"""

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel
from typing import Optional, List, Dict, Any

from services.syntax_service import get_syntax_service

router = APIRouter(prefix="/api/syntax", tags=["syntax"])


class SyntaxRequest(BaseModel):
    """Request model for syntax analysis"""
    sentence: str
    language: str = "english"


class DependencyRequest(BaseModel):
    """Request model for dependency parsing with options"""
    sentence: str
    language: str = "english"
    compact: bool = False
    collapse_punct: bool = True
    collapse_phrases: bool = False


class ConstituencyResponse(BaseModel):
    """Response model for constituency parsing"""
    success: bool
    tree_string: str = ""
    tree_data: Optional[Dict[str, Any]] = None
    sentence: str = ""
    error: Optional[str] = None


class DependencyResponse(BaseModel):
    """Response model for dependency parsing"""
    success: bool
    svg_html: str = ""
    tokens: List[Dict[str, Any]] = []
    arcs: List[Dict[str, Any]] = []
    sentence: str = ""
    error: Optional[str] = None


class StatusResponse(BaseModel):
    """Response model for service status"""
    constituency_available: bool
    dependency_available: bool
    dependency_languages: List[str]


class LabelsResponse(BaseModel):
    """Response model for label descriptions"""
    labels: Dict[str, str]


@router.post("/constituency", response_model=ConstituencyResponse)
async def analyze_constituency(request: SyntaxRequest):
    """
    Perform constituency parsing on a sentence
    
    Returns parse tree in both string and hierarchical data format
    """
    service = get_syntax_service()
    
    if not request.sentence.strip():
        raise HTTPException(status_code=400, detail="Sentence cannot be empty")
    
    result = service.analyze_constituency(
        sentence=request.sentence.strip(),
        language=request.language
    )
    
    return ConstituencyResponse(**result)


@router.post("/dependency", response_model=DependencyResponse)
async def analyze_dependency(request: DependencyRequest):
    """
    Perform dependency parsing on a sentence
    
    Returns SVG visualization and token/arc information
    Options:
    - compact: Use straight lines instead of arcs
    - collapse_punct: Collapse punctuation
    - collapse_phrases: Collapse phrases
    """
    service = get_syntax_service()
    
    if not request.sentence.strip():
        raise HTTPException(status_code=400, detail="Sentence cannot be empty")
    
    result = service.analyze_dependency(
        sentence=request.sentence.strip(),
        language=request.language,
        compact=request.compact,
        collapse_punct=request.collapse_punct,
        collapse_phrases=request.collapse_phrases
    )
    
    return DependencyResponse(**result)


@router.get("/status", response_model=StatusResponse)
async def get_status():
    """
    Get syntax analysis service status
    
    Returns availability of constituency and dependency parsing
    """
    service = get_syntax_service()
    
    return StatusResponse(
        constituency_available=service.is_constituency_available(),
        dependency_available=service.is_dependency_available("english"),
        dependency_languages=["english", "chinese"]
    )


@router.get("/labels/dependency", response_model=LabelsResponse)
async def get_dependency_labels():
    """
    Get dependency relation label descriptions
    """
    service = get_syntax_service()
    return LabelsResponse(labels=service.get_dependency_labels())


@router.get("/labels/constituency", response_model=LabelsResponse)
async def get_constituency_labels():
    """
    Get constituency phrase label descriptions
    """
    service = get_syntax_service()
    return LabelsResponse(labels=service.get_constituency_labels())

File: src/api/syntax.ts
/**
 * Syntax Analysis API
 * Client for constituency and dependency parsing endpoints
 */

import { api } from './client'
import type { ApiResponse } from '../types'

// Types for syntax analysis
export interface SyntaxRequest {
  sentence: string
  language?: string
}

export interface DependencyOptions {
  compact?: boolean
  collapse_punct?: boolean
  collapse_phrases?: boolean
}

export interface TreeNode {
  label: string
  children: TreeNode[]
  text: string
  isLeaf?: boolean
}

export interface ConstituencyResponse {
  success: boolean
  tree_string: string
  tree_data: TreeNode | null
  sentence: string
  error: string | null
}

export interface DependencyToken {
  id: number
  text: string
  lemma: string
  pos: string
  tag: string
  dep: string
  head_id: number
  head_text: string
}

export interface DependencyArc {
  start: number
  end: number
  label: string
  dir: 'left' | 'right'
}

export interface DependencyResponse {
  success: boolean
  svg_html: string
  tokens: DependencyToken[]
  arcs: DependencyArc[]
  sentence: string
  error: string | null
}

export interface SyntaxStatus {
  constituency_available: boolean
  dependency_available: boolean
  dependency_languages: string[]
}

export interface LabelDescriptions {
  labels: Record<string, string>
}

/**
 * Syntax Analysis API client
 */
export const syntaxApi = {
  /**
   * Analyze constituency structure of a sentence
   */
  analyzeConstituency: async (
    sentence: string,
    language: string = 'english'
  ): Promise<ApiResponse<ConstituencyResponse>> => {
    return api.post<ConstituencyResponse>('/api/syntax/constituency', {
      sentence,
      language
    })
  },

  /**
   * Analyze dependency structure of a sentence
   */
  analyzeDependency: async (
    sentence: string,
    language: string = 'english',
    options?: DependencyOptions
  ): Promise<ApiResponse<DependencyResponse>> => {
    return api.post<DependencyResponse>('/api/syntax/dependency', {
      sentence,
      language,
      compact: options?.compact ?? false,
      collapse_punct: options?.collapse_punct ?? true,
      collapse_phrases: options?.collapse_phrases ?? false
    })
  },

  /**
   * Get syntax service status
   */
  getStatus: async (): Promise<ApiResponse<SyntaxStatus>> => {
    return api.get<SyntaxStatus>('/api/syntax/status')
  },

  /**
   * Get dependency label descriptions
   */
  getDependencyLabels: async (): Promise<ApiResponse<LabelDescriptions>> => {
    return api.get<LabelDescriptions>('/api/syntax/labels/dependency')
  },

  /**
   * Get constituency label descriptions
   */
  getConstituencyLabels: async (): Promise<ApiResponse<LabelDescriptions>> => {
    return api.get<LabelDescriptions>('/api/syntax/labels/constituency')
  }
}

export default syntaxApi
